File: admin-api/src/main/java/me/synology/hajubal/coins/admin/service/DashboardService.java
package me.synology.hajubal.coins.admin.service;

import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import me.synology.hajubal.coins.admin.dto.DashboardStatsDto;
import me.synology.hajubal.coins.entity.SavedPoint;
import me.synology.hajubal.coins.respository.PointUrlRepository;
import me.synology.hajubal.coins.respository.SavedPointRepository;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDate;
import java.time.LocalDateTime;
import java.time.LocalTime;
import java.util.List;

@Slf4j
@Service
@RequiredArgsConstructor
@Transactional(readOnly = true)
public class DashboardService {

  private final SavedPointRepository savedPointRepository;
  private final PointUrlRepository pointUrlRepository;

  public DashboardStatsDto getStats() {
    // 오늘 날짜 범위
    LocalDateTime todayStart = LocalDateTime.of(LocalDate.now(), LocalTime.MIN);
    LocalDateTime todayEnd = LocalDateTime.of(LocalDate.now(), LocalTime.MAX);

    // 어제 날짜 범위
    LocalDateTime yesterdayStart = LocalDateTime.of(LocalDate.now().minusDays(1), LocalTime.MIN);
    LocalDateTime yesterdayEnd = LocalDateTime.of(LocalDate.now().minusDays(1), LocalTime.MAX);

    // 이번 주 날짜 범위 (월요일부터)
    LocalDate today = LocalDate.now();
    LocalDate weekStart = today.minusDays(today.getDayOfWeek().getValue() - 1);
    LocalDateTime thisWeekStart = LocalDateTime.of(weekStart, LocalTime.MIN);
    LocalDateTime thisWeekEnd = todayEnd;

    // 지난 주 날짜 범위
    LocalDate lastWeekStart = weekStart.minusWeeks(1);
    LocalDate lastWeekEnd = weekStart.minusDays(1);
    LocalDateTime lastWeekStartTime = LocalDateTime.of(lastWeekStart, LocalTime.MIN);
    LocalDateTime lastWeekEndTime = LocalDateTime.of(lastWeekEnd, LocalTime.MAX);

    // 오늘 포인트 합계
    List<SavedPoint> todayPoints = savedPointRepository.findAllByCreatedDateBetween(todayStart, todayEnd);
    int savedDayPoint = todayPoints.stream().mapToInt(SavedPoint::getAmount).sum();

    // 어제 포인트 합계
    List<SavedPoint> yesterdayPoints =
        savedPointRepository.findAllByCreatedDateBetween(yesterdayStart, yesterdayEnd);
    int yesterdaySavedPoint = yesterdayPoints.stream().mapToInt(SavedPoint::getAmount).sum();

    // 전일 대비 비율 계산
    double savedDayPointRatioDayBefore = 0.0;
    if (yesterdaySavedPoint > 0) {
      savedDayPointRatioDayBefore =
          ((double) (savedDayPoint - yesterdaySavedPoint) / yesterdaySavedPoint) * 100;
    } else if (savedDayPoint > 0) {
      savedDayPointRatioDayBefore = 100.0;
    }

    // 이번 주 포인트 합계
    List<SavedPoint> thisWeekPoints =
        savedPointRepository.findAllByCreatedDateBetween(thisWeekStart, thisWeekEnd);
    int savedWeekPoint = thisWeekPoints.stream().mapToInt(SavedPoint::getAmount).sum();

    // 지난 주 포인트 합계
    List<SavedPoint> lastWeekPoints =
        savedPointRepository.findAllByCreatedDateBetween(lastWeekStartTime, lastWeekEndTime);
    int lastWeekSavedPoint = lastWeekPoints.stream().mapToInt(SavedPoint::getAmount).sum();

    // 전주 대비 비율 계산
    double savedWeekPointRatioWeekBefore = 0.0;
    if (lastWeekSavedPoint > 0) {
      savedWeekPointRatioWeekBefore =
          ((double) (savedWeekPoint - lastWeekSavedPoint) / lastWeekSavedPoint) * 100;
    } else if (savedWeekPoint > 0) {
      savedWeekPointRatioWeekBefore = 100.0;
    }

    // 오늘 생성된 PointUrl 수
    long pointUrlDayCnt = pointUrlRepository.findByCreatedDateBetween(todayStart, todayEnd).size();

    // 이번 주 생성된 PointUrl 수
    long pointUrlWeekCnt =
        pointUrlRepository.findByCreatedDateBetween(thisWeekStart, thisWeekEnd).size();

    log.info(
        "Dashboard stats - Today points: {}, Week points: {}, Today URLs: {}, Week URLs: {}",
        savedDayPoint,
        savedWeekPoint,
        pointUrlDayCnt,
        pointUrlWeekCnt);

    return DashboardStatsDto.builder()
        .savedDayPoint(savedDayPoint)
        .savedDayPointRatioDayBefore(Math.round(savedDayPointRatioDayBefore * 10.0) / 10.0)
        .savedWeekPoint(savedWeekPoint)
        .savedWeekPointRatioWeekBefore(Math.round(savedWeekPointRatioWeekBefore * 10.0) / 10.0)
        .pointUrlDayCnt(pointUrlDayCnt)
        .pointUrlWeekCnt(pointUrlWeekCnt)
        .build();
  }
}

File: src/modules/admin/dashboard/dashboard.service.ts
import { Injectable, Logger } from '@nestjs/common';
import { PrismaService } from '../../prisma/prisma.service';
import { DashboardStatsDto } from './dto/dashboard.dto';

@Injectable()
export class DashboardService {
  private readonly logger = new Logger(DashboardService.name);

  constructor(private readonly prisma: PrismaService) {}

  async getStats(): Promise<DashboardStatsDto> {
    const now = new Date();
    const today = new Date(now.getFullYear(), now.getMonth(), now.getDate());
    const tomorrow = new Date(today);
    tomorrow.setDate(tomorrow.getDate() + 1);

    // Yesterday range
    const yesterdayStart = new Date(today);
    yesterdayStart.setDate(yesterdayStart.getDate() - 1);
    const yesterdayEnd = new Date(today);

    // This week range (from Monday)
    const dayOfWeek = today.getDay();
    const mondayOffset = dayOfWeek === 0 ? -6 : 1 - dayOfWeek;
    const thisWeekStart = new Date(today);
    thisWeekStart.setDate(thisWeekStart.getDate() + mondayOffset);
    const thisWeekEnd = tomorrow;

    // Last week range
    const lastWeekStart = new Date(thisWeekStart);
    lastWeekStart.setDate(lastWeekStart.getDate() - 7);
    const lastWeekEnd = new Date(thisWeekStart);

    // Today's points
    const todayPointsResult = await this.prisma.savedPoint.aggregate({
      where: {
        createdDate: { gte: today, lt: tomorrow },
      },
      _sum: { amount: true },
    });
    const savedDayPoint = todayPointsResult._sum.amount || 0;

    // Yesterday's points
    const yesterdayPointsResult = await this.prisma.savedPoint.aggregate({
      where: {
        createdDate: { gte: yesterdayStart, lt: yesterdayEnd },
      },
      _sum: { amount: true },
    });
    const yesterdaySavedPoint = yesterdayPointsResult._sum.amount || 0;

    // Day-over-day ratio
    let savedDayPointRatioDayBefore = 0;
    if (yesterdaySavedPoint > 0) {
      savedDayPointRatioDayBefore = ((savedDayPoint - yesterdaySavedPoint) / yesterdaySavedPoint) * 100;
    } else if (savedDayPoint > 0) {
      savedDayPointRatioDayBefore = 100;
    }

    // This week's points
    const thisWeekPointsResult = await this.prisma.savedPoint.aggregate({
      where: {
        createdDate: { gte: thisWeekStart, lt: thisWeekEnd },
      },
      _sum: { amount: true },
    });
    const savedWeekPoint = thisWeekPointsResult._sum.amount || 0;

    // Last week's points
    const lastWeekPointsResult = await this.prisma.savedPoint.aggregate({
      where: {
        createdDate: { gte: lastWeekStart, lt: lastWeekEnd },
      },
      _sum: { amount: true },
    });
    const lastWeekSavedPoint = lastWeekPointsResult._sum.amount || 0;

    // Week-over-week ratio
    let savedWeekPointRatioWeekBefore = 0;
    if (lastWeekSavedPoint > 0) {
      savedWeekPointRatioWeekBefore = ((savedWeekPoint - lastWeekSavedPoint) / lastWeekSavedPoint) * 100;
    } else if (savedWeekPoint > 0) {
      savedWeekPointRatioWeekBefore = 100;
    }

    // Today's point URLs count
    const pointUrlDayCnt = await this.prisma.pointUrl.count({
      where: {
        createdDate: { gte: today, lt: tomorrow },
      },
    });

    // This week's point URLs count
    const pointUrlWeekCnt = await this.prisma.pointUrl.count({
      where: {
        createdDate: { gte: thisWeekStart, lt: thisWeekEnd },
      },
    });

    this.logger.log(
      `Dashboard stats - Today points: ${savedDayPoint}, Week points: ${savedWeekPoint}, Today URLs: ${pointUrlDayCnt}, Week URLs: ${pointUrlWeekCnt}`,
    );

    return {
      savedDayPoint,
      savedDayPointRatioDayBefore: Math.round(savedDayPointRatioDayBefore * 10) / 10,
      savedWeekPoint,
      savedWeekPointRatioWeekBefore: Math.round(savedWeekPointRatioWeekBefore * 10) / 10,
      pointUrlDayCnt,
      pointUrlWeekCnt,
    };
  }
}
